runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/BundleCheckpointHandlers.java

```diff
@@ -0,0 +1,138 @@
+    private final TimerInternalsFactory<T> timerInternalsFactory;
+    private final StateInternalsFactory<T> stateInternalsFactory;
+    private final Coder<WindowedValue<T>> residualCoder;
+    private final Coder windowCoder;
+    private final IdGenerator idGenerator = IdGenerators.incrementingLongs();
+    public static final String SDF_PREFIX = "sdf_checkpoint";
+
+    public StateAndTimerBundleCheckpointHandler(
+        TimerInternalsFactory<T> timerInternalsFactory,
+        StateInternalsFactory<T> stateInternalsFactory,
+        Coder<WindowedValue<T>> residualCoder,
+        Coder windowCoder) {
+      this.residualCoder = residualCoder;
+      this.windowCoder = windowCoder;
+      this.timerInternalsFactory = timerInternalsFactory;
+      this.stateInternalsFactory = stateInternalsFactory;
+    }
+
+    /**
+     * A helper function to help check whether the given timer is the timer which is set for
+     * rescheduling {@link DelayedBundleApplication}.
+     */
+    public static boolean isSdfTimer(String timerId) {
+      return timerId.startsWith(SDF_PREFIX);
+    }
+
+    private static String constructSdfCheckpointId(String id, int index) {
+      return SDF_PREFIX + ":" + id + ":" + index;
+    }
+
+    @Override
+    public void onCheckpoint(ProcessBundleResponse response) {
+      String id = idGenerator.getId();
+      for (int index = 0; index < response.getResidualRootsCount(); index++) {
+        DelayedBundleApplication residual = response.getResidualRoots(index);
+        if (!residual.hasApplication()) {
+          continue;
+        }
+        String tag = constructSdfCheckpointId(id, index);
+        try {
+          WindowedValue<T> stateValue =
+              CoderUtils.decodeFromByteArray(
+                  residualCoder, residual.getApplication().getElement().toByteArray());
+          TimerInternals timerInternals =
+              timerInternalsFactory.timerInternalsForKey((stateValue.getValue()));
+          StateInternals stateInternals =
+              stateInternalsFactory.stateInternalsForKey(stateValue.getValue());
+          // Calculate the timestamp for the timer.
+          Instant timestamp = Instant.now();
+          if (residual.hasRequestedTimeDelay()) {
+            timestamp = timestamp.plus(residual.getRequestedTimeDelay().getSeconds() * 1000);
+          }
+          // Calculate the watermark hold for the timer.
+          long outputTimestamp = BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis();
+          if (!residual.getApplication().getOutputWatermarksMap().isEmpty()) {
+            for (org.apache.beam.vendor.grpc.v1p26p0.com.google.protobuf.Timestamp outputWatermark :
+                residual.getApplication().getOutputWatermarksMap().values()) {
+              outputTimestamp = Math.min(outputTimestamp, outputWatermark.getSeconds() * 1000);
+            }
+          } else {
+            outputTimestamp = BoundedWindow.TIMESTAMP_MIN_VALUE.getMillis();
```

Comment:
Is it a good idea to rely on `TimerInternals.getCurrentInputWatermark()` to not hold back the watermark so much?
@mxm

Reply:
You could use `keyedStateInternals.minWatermarkHoldMs()`.

Reply:
Thanks, Max! I just double checked the contract we have: https://github.com/apache/beam/blob/7eb0e956dad87f84622d3523369579812ff82f0a/model/fn-execution/src/main/proto/beam_fn_api.proto#L236-L243
Let's stick with MIN_TIMESTAMP for now.